packages/test/test-end-to-end-tests/src/test/orderSequentially.spec.ts

```diff
@@ -0,0 +1,218 @@
+
+		assert.equal(changedEventData[2], 2);
+
+		// rollback
+		assert.equal(changedEventData[3].key, "key1");
+		assert.equal(changedEventData[3].previousValue, 0);
+
+		assert.equal(changedEventData[4].key, "key2");
+		assert.equal(changedEventData[4].previousValue, 1);
+
+		assert.equal(changedEventData[5], undefined);
+
+		assert.equal(changedEventData[6], 2);
+
+		assert.equal(changedEventData[7].key, "key2");
+		assert.equal(changedEventData[7].previousValue, undefined);
+
+		assert.equal(changedEventData[8].key, "key1");
+		assert.equal(changedEventData[8].previousValue, undefined);
+	});
+
+	it("Should rollback complex edits on multiple DDS types", () => {
+		sharedString.insertText(0, "abcde");
+		sharedMap.set("key1", 0);
+		sharedString.annotateRange(2, 5, { foo: "old" });
+		sharedDir.set("key2", 1);
+		sharedCell.set(2);
+		try {
+			containerRuntime.orderSequentially(() => {
+				sharedMap.set("key1", 3);
+				sharedString.annotateRange(0, 3, { foo: "new" });
+				sharedDir.delete("key2");
+				sharedCell.set(5);
+				sharedString.removeRange(0, 5);
+				sharedCell.delete();
+				throw new Error(errorMessage);
+			});
+		} catch (err) {
+			error = err as Error;
+		}
+
+		assert.notEqual(error, undefined, "No error");
+		assert.equal(error?.message, errorMessage, "Unexpected error message");
+		assert.equal(containerRuntime.disposed, false, "Container disposed");
+		assert.equal(sharedString.getText(), "abcde");
+		for (let i = 0; i < sharedString.getText().length; i++) {
+			const props = sharedString.getPropertiesAtPosition(i);
+			if (i >= 2 && i < 5) {
+				assert.equal(props?.foo, "old");
+			} else {
+				assert(props === undefined || props.foo === undefined);
+			}
+		}
+		assert.equal(sharedMap.size, 1);
+		assert.equal(sharedMap.has("key1"), true);
+		assert.equal(sharedMap.get("key1"), 0);
+		assert.equal(sharedDir.size, 1);
+		assert.equal(sharedDir.has("key2"), true);
+		assert.equal(sharedDir.get("key2"), 1);
+		assert.equal(sharedCell.get(), 2);
+
+		assert.equal(changedEventData.length, 11);
+		assert.equal(changedEventData[0].key, "key1");
+		assert.equal(changedEventData[0].previousValue, undefined);
+
+		assert.equal(changedEventData[1].key, "key2");
+		assert.equal(changedEventData[1].previousValue, undefined);
+
+		assert.equal(changedEventData[2], 2);
+
+		// rollback
+		assert.equal(changedEventData[3].key, "key1");
+		assert.equal(changedEventData[3].previousValue, 0);
+
+		assert.equal(changedEventData[4].key, "key2");
+		assert.equal(changedEventData[4].previousValue, 1);
+
+		assert.equal(changedEventData[5], 5);
+
+		assert.equal(changedEventData[6], undefined);
+
+		assert.equal(changedEventData[7], 5);
+
+		assert.equal(changedEventData[8], 2);
+
+		assert.equal(changedEventData[9].key, "key2");
+		assert.equal(changedEventData[9].previousValue, undefined);
+
+		assert.equal(changedEventData[10].key, "key1");
+		assert.equal(changedEventData[10].previousValue, 3);
+	});
+});
```

Comment:
orderSequentially can be nested. It would be good to have [relatively simple] tests where the nested rollback is invoked and the outer rollback is invoked after the nested one is run. Doesn't have to be in this PR.